libpages/server.go

```diff
@@ -121,12 +137,64 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			msg:    "CtxWithRandomIDReplayable",
 			logger: s.config.Logger,
 		})
-	fs, err := s.getFS(ctx, root)
+	st, err := s.getSite(ctx, root)
+	if err != nil {
+		s.handleError(w, err)
+		return
+	}
+
+	cfg, err := st.getConfig(false)
+	if err != nil {
+		// TODO: error page to show the error message?
+		s.config.Logger.Info("getConfig", zap.String("host", r.Host), zap.Error(err))
+		s.handleError(w, err)
+		return
+	}
+
+	var canRead, canList bool
+	var realm string
+	user, pass, ok := r.BasicAuth()
+	if ok && cfg.Authenticate(user, pass) {
+		canRead, canList, realm, err = cfg.GetPermissionsForUsername(
+			r.RequestURI, user)
+	} else {
+		canRead, canList, realm, err = cfg.GetPermissionsForAnonymous(
+			r.RequestURI)
+	}
 	if err != nil {
 		s.handleError(w, err)
 		return
 	}
-	http.FileServer(fs).ServeHTTP(w, r)
+
+	if !canRead {
+		s.handleNeedAuthentication(w, r, realm)
```

Comment:
If it's a directory with no index.html file, shouldn't we allow listing if `canList == true`?  Or do you intend for reading to be a requirement for listing?

Reply:
I thought `read` as a prerequisite for `list`, as it'd be like reading a directory. But I guess there's a use case here where one can have `list` on `/foo` which has a bunch of sub-directories, only some of which have `read` for a particular user. Will update to apply your logic!